java-checks/src/main/java/org/sonar/java/checks/RedundantModifierCheck.java

```diff
@@ -47,35 +47,71 @@
   public void visitNode(Tree tree) {
     ClassTree classTree = (ClassTree) tree;
     for (Tree member : classTree.members()) {
-      if (member.is(Tree.Kind.METHOD)) {
-        MethodTree methodTree = (MethodTree) member;
-        ModifiersTree modifiers = methodTree.modifiers();
-        if (isInterfaceOrAnnotation(tree)) {
-          checkRedundantModifier(modifiers, Modifier.ABSTRACT);
-          checkRedundantModifier(modifiers, Modifier.PUBLIC);
-        } else if (ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
-          checkRedundantModifier(modifiers, Modifier.FINAL);
-        }
-      } else if (member.is(Tree.Kind.VARIABLE) && isInterfaceOrAnnotation(tree)) {
-        VariableTree variableTree = (VariableTree) member;
-        ModifiersTree modifiers = variableTree.modifiers();
-        checkRedundantModifier(modifiers, Modifier.PUBLIC);
-        checkRedundantModifier(modifiers, Modifier.STATIC);
-        checkRedundantModifier(modifiers, Modifier.FINAL);
-      } else if(member.is(Kind.CONSTRUCTOR) && tree.is(Kind.ENUM)) {
-        checkRedundantModifier(((MethodTree) member).modifiers(), Modifier.PRIVATE);
+      switch (member.kind()) {
+        case METHOD:
+          checkMethod((MethodTree) member, classTree);
+          break;
+        case VARIABLE:
+          checkVariable((VariableTree) member, classTree);
+          break;
+        case CONSTRUCTOR:
+          if (tree.is(Kind.ENUM)) {
+            checkRedundantModifier(((MethodTree) member).modifiers(), Modifier.PRIVATE);
+          }
+          break;
+        case INTERFACE:
+          ClassTree nestedClass = (ClassTree) member;
+          checkNestedInterface(nestedClass, classTree);
+          checkNestedClass(nestedClass, classTree);
+          break;
+        case CLASS:
+          checkNestedClass((ClassTree) member, classTree);
+          break;
+        default:
+          // Do nothing for others members
       }
     }
   }
 
+  private void checkMethod(MethodTree methodTree, ClassTree classTree) {
+    ModifiersTree modifiers = methodTree.modifiers();
+    if (isInterfaceOrAnnotation(classTree)) {
+      checkRedundantModifier(modifiers, Modifier.ABSTRACT, Modifier.PUBLIC);
+    } else if (ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
+      checkRedundantModifier(modifiers, Modifier.FINAL);
+    }
+  }
+
+  private void checkVariable(VariableTree variableTree, ClassTree classTree) {
+    if (isInterfaceOrAnnotation(classTree)) {
+      ModifiersTree modifiers = variableTree.modifiers();
+      checkRedundantModifier(modifiers, Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL);
+    }
+  }
+
+  private void checkNestedClass(ClassTree nested, ClassTree classTree) {
```

Comment:
Maybe it would be nice to rename it to sth more general like `checkNestedType`. Cause this method is used on both classes and interfaces, so a little bit confusing.